pkg/plugins/runtime/k8s/controllers/gatewayapi/gateway_controller.go

```diff
@@ -0,0 +1,198 @@
+	mesh := k8s_util.MeshFor(gateway)
+
+	resource.Spec = r.gapiToKumaGateway(gateway)
+
+	if err := r.ResourceManager.Get(ctx, resource, store.GetByKey(coreName, mesh)); err != nil {
+		if !store.IsResourceNotFound(err) {
+			return kube_ctrl.Result{}, errors.Wrap(err, "error retrieving Gateway from Kuma store")
+		}
+
+		resource.SetMeta(&k8s.KubernetesMetaAdapter{ObjectMeta: metav1.ObjectMeta{
+			Name:      gateway.Name,
+			Namespace: gateway.Namespace,
+		}, Mesh: mesh})
+
+		err := r.ResourceManager.Create(ctx, resource, store.CreateByKey(coreName, mesh))
+		if err != nil {
+			return kube_ctrl.Result{}, errors.Wrap(err, "could not create Gateway resource")
+		}
+	} else if err := r.ResourceManager.Update(ctx, resource, store.ModifiedAt(core.Now())); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "could not update Gateway resource")
+	}
+
+	deployment, err := r.createOrUpdateDeployment(ctx, resource, gateway)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to create Deployment for Gateway")
+	}
+
+	svc, err := r.createOrUpdateService(ctx, resource, gateway)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to create Service for Gateway")
+	}
+
+	r.updateStatus(gateway, svc, deployment)
+
+	if err := r.Client.Status().Update(ctx, gateway); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to update Gateway status")
+	}
+
+	return kube_ctrl.Result{}, nil
+}
+
+func (r *GatewayReconciler) getGatewayClass(ctx context.Context, name gatewayapi.ObjectName) (*gatewayapi.GatewayClass, error) {
+	class := &gatewayapi.GatewayClass{}
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Name: string(name)}, class); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return nil, nil
+		}
+
+		return nil, errors.Wrap(err, "unable to lookup GatewayClass")
+	}
+
+	if class.Spec.ControllerName != controllerName {
+		return nil, nil
+	}
+
+	return class, nil
+}
+
+func (r *GatewayReconciler) gapiToKumaGateway(gateway *gatewayapi.Gateway) *mesh_proto.Gateway {
+	var listeners []*mesh_proto.Gateway_Listener
+
+	for _, l := range gateway.Spec.Listeners {
+		listener := &mesh_proto.Gateway_Listener{
+			Port: uint32(l.Port),
+			Tags: map[string]string{
+				// gateway-api routes are configured using direct references to
+				// Gateways, so just create a tag specifically for this listener
+				mesh_proto.ListenerTag: string(l.Name),
+			},
+		}
+
+		if protocol, ok := mesh_proto.Gateway_Listener_Protocol_value[string(l.Protocol)]; ok {
+			listener.Protocol = mesh_proto.Gateway_Listener_Protocol(protocol)
+		}
+
+		if hn := l.Hostname; hn != nil {
+			listener.Hostname = string(*hn)
+		}
+
+		listeners = append(listeners, listener)
+	}
+
+	match := map[string]string{
+		mesh_proto.ServiceTag: serviceTagForGateway(gateway.Name, gateway.Namespace),
+	}
+	if m := k8s_util.MeshFor(gateway); m != model.DefaultMesh {
+		match[metadata.KumaMeshAnnotation] = m
+	}
```

Comment:
Since we are generating these, I'd recommend erring on the side of being more explicit, so adding the mesh tag to everything even if it is the default mesh seems fine. OK to leave a TODO and address a separate PR if you prefer.

Reply:
yep, doesn't make sense. gone